### ISSUE TITLE
False positive for overload parameters overlapping with specific recursive types

### ISSUE THREAD

**Issue by Kajiih** [microsoft/pyright#9514]
**Describe the bug**
There seems to be a false positive for overload parameters overlapping, happening specifically with some recursive types.

I couldn't point out the exact issue, so here's an example:
**Code or Screenshots**
It is not really a minimal example, but I tried to keep some context. In particular, the protocol `CallableArbitraryNbArgs` was originally generic, but the false positive happens in both generic and non-generic cases.

The false positive happens when using `NestedMapping` which is a recursive type that involves `NestedMappingNode`, but not when using `NestedMapping2`, that doesn't.

```python
from collections.abc import Callable, Mapping
from typing import Any, Protocol, overload

type NestedMapping[K, V] = Mapping[K, NestedMappingNode[K, V]]
type NestedMappingNode[K, V] = V | NestedMapping[K, V]


type NestedMapping2[K, V] = Mapping[K, NestedMapping2[K, V]]


class CallableArbitraryNbArgs(Protocol):
    """Protocol for a callable that accepts an arbitrary number of arguments."""

    def __call__(self, *args: Any) -> Any: ...  # noqa: D102


# === With NestedMapping: false positive ===
@overload
def map_leaves_1[K, V, W](
    func: CallableArbitraryNbArgs,
    *nested_dicts: NestedMapping[K, V],
) -> NestedMapping[K, W]: ...


# Overload 2 for "map_leaves_no_generic_protocol" will never be used because its parameters overlap overload 1
@overload
def map_leaves_1[K, W](
    func: Callable[..., W],
    *nested_dicts: NestedMapping[K, Any],
) -> NestedMapping[K, W]: ...


def map_leaves_1[K, W](
    func: Callable[..., W] | CallableArbitraryNbArgs, *nested_dicts: NestedMapping[K, Any]
) -> NestedMapping[K, W]:
    del nested_dicts
    del func
    raise NotImplementedError


# === With NestedMapping2: no error ===
@overload
def map_leaves_2[K, V, W](
    func: CallableArbitraryNbArgs,
    *nested_dicts: NestedMapping2[K, V],
) -> NestedMapping[K, W]: ...


@overload
def map_leaves_2[K, W](
    func: Callable[..., W],
    *nested_dicts: NestedMapping2[K, Any],
) -> NestedMapping[K, W]: ...


def map_leaves_2[K, W](
    func: Callable[..., W] | CallableArbitraryNbArgs, *nested_dicts: NestedMapping2[K, Any]
) -> NestedMapping[K, W]:
    del nested_dicts
    del func
    raise NotImplementedError
```
If I'm not mistaking, the overload parameters don't overlap in either case.

**VS Code extension or command-line**
Both extension and command line, Pylance and Pyright version `1.1.389`

**Comment by loing-cdut** [comment]
You've encountered an issue with type annotations, particularly when dealing with recursive types and overloaded functions. This issue involves Python's type hinting system, specifically the use of the typing module and Protocol.

Here's the analysis of your code:

You defined two type aliases, NestedMapping and NestedMapping2, which are recursive and represent nested mapping types.

You defined a CallableArbitraryNbArgs protocol, which represents a callable that accepts an arbitrary number of arguments.

You attempted to define overloads for the map_leaves_1 and map_leaves_2 functions, which accept a function and multiple nested dictionaries as arguments and return a nested dictionary.

The issue arises with the overload definitions of map_leaves_1. The second overload map_leaves_1[K, W] has overlapping parameters with the first overload map_leaves_1[K, V, W], leading to a false positive. This is because in the second overload, you used NestedMapping[K, Any] as the parameter type, which is compatible with NestedMapping[K, V] from the first overload, since Any is a supertype of all types. This means the second overload can be seen as a more general version of the first overload, violating the rules of Python overloads, where each overload should be mutually exclusive and not overlap in parameters.

To resolve this issue, you need to ensure that the parameters of each overload are mutually exclusive and do not overlap. You can address this by changing the parameter type of the second overload, for example, by using a different type alias or adding additional constraints to differentiate between the two overloads.

Here's an example of the modified code in English:

python
from collections.abc import Callable, Mapping
from typing import Any, Protocol, overload

# Define a recursive type alias for nested mappings
NestedMapping = Mapping[K, NestedMappingNode[K, V]]
NestedMappingNode = V | Mapping[K, NestedMappingNode[K, V]]

# Define a non-generic nested mapping type alias
NonGenericNestedMapping = Mapping[K, Any]

# Define a protocol for a callable that accepts an arbitrary number of arguments
class CallableArbitraryNbArgs(Protocol):
    def __call__(self, *args: Any) -> Any: ...

# Overload 1 for map_leaves_1 with generic protocol
@overload
def map_leaves_1[K, V, W](
    func: CallableArbitraryNbArgs,
    *nested_dicts: NestedMapping[K, V],
) -> NestedMapping[K, W]: ...

# Overload 2 for map_leaves_1 with non-generic protocol
@overload
def map_leaves_1[K, W](
    func: Callable[..., W],
    *nested_dicts: NonGenericNestedMapping[K],
) -> NestedMapping[K, W]: ...

# Implementation of map_leaves_1 with a union of callable types and non-generic nested mapping
def map_leaves_1[K, W](
    func: (Callable[..., W] | CallableArbitraryNbArgs), *nested_dicts: Mapping[K, Any]
) -> NestedMapping[K, W]:
    del nested_dicts
    del func
    raise NotImplementedError
In this modification, I created a new type alias NonGenericNestedMapping, which does not contain generic parameters, ensuring that the second overload's parameter type does not overlap with the first overload's parameter type. This way, your overload definitions should not produce false positives.

**Comment by loing-cdut** [comment]
You've encountered an issue with type annotations, particularly when dealing with recursive types and overloaded functions. This issue involves Python's type hinting system, specifically the use of the typing module and Protocol.

Here's the analysis of your code:

You defined two type aliases, NestedMapping and NestedMapping2, which are recursive and represent nested mapping types.

You defined a CallableArbitraryNbArgs protocol, which represents a callable that accepts an arbitrary number of arguments.

You attempted to define overloads for the map_leaves_1 and map_leaves_2 functions, which accept a function and multiple nested dictionaries as arguments and return a nested dictionary.

The issue arises with the overload definitions of map_leaves_1. The second overload map_leaves_1[K, W] has overlapping parameters with the first overload map_leaves_1[K, V, W], leading to a false positive. This is because in the second overload, you used NestedMapping[K, Any] as the parameter type, which is compatible with NestedMapping[K, V] from the first overload, since Any is a supertype of all types. This means the second overload can be seen as a more general version of the first overload, violating the rules of Python overloads, where each overload should be mutually exclusive and not overlap in parameters.

To resolve this issue, you need to ensure that the parameters of each overload are mutually exclusive and do not overlap. You can address this by changing the parameter type of the second overload, for example, by using a different type alias or adding additional constraints to differentiate between the two overloads.

Here's an example of the modified code in English:

python
from collections.abc import Callable, Mapping
from typing import Any, Protocol, overload

# Define a recursive type alias for nested mappings
NestedMapping = Mapping[K, NestedMappingNode[K, V]]
NestedMappingNode = V | Mapping[K, NestedMappingNode[K, V]]

# Define a non-generic nested mapping type alias
NonGenericNestedMapping = Mapping[K, Any]

# Define a protocol for a callable that accepts an arbitrary number of arguments
class CallableArbitraryNbArgs(Protocol):
    def __call__(self, *args: Any) -> Any: ...

# Overload 1 for map_leaves_1 with generic protocol
@overload
def map_leaves_1[K, V, W](
    func: CallableArbitraryNbArgs,
    *nested_dicts: NestedMapping[K, V],
) -> NestedMapping[K, W]: ...

# Overload 2 for map_leaves_1 with non-generic protocol
@overload
def map_leaves_1[K, W](
    func: Callable[..., W],
    *nested_dicts: NonGenericNestedMapping[K],
) -> NestedMapping[K, W]: ...

# Implementation of map_leaves_1 with a union of callable types and non-generic nested mapping
def map_leaves_1[K, W](
    func: (Callable[..., W] | CallableArbitraryNbArgs), *nested_dicts: Mapping[K, Any]
) -> NestedMapping[K, W]:
    del nested_dicts
    del func
    raise NotImplementedError
In this modification, I created a new type alias NonGenericNestedMapping, which does not contain generic parameters, ensuring that the second overload's parameter type does not overlap with the first overload's parameter type. This way, your overload definitions should not produce false positives.

**Comment by erictraut** [comment]
Thanks for the bug report. I agree that this should not be considered a fully-overlapping override, but perhaps not for the reason you think. It has nothing to do with recursive types. It's because `CallableArbitraryNbArgs` does not accept keyword arguments.

Here's a code sample that demonstrates how a callback function with keyword parameters would not be accepted by the first overload but would be accepted by the second.

```python
from typing import Any, Callable, Protocol

class CB(Protocol):
    def __call__(self, *args: Any) -> Any: ...

def overload1(func: CB) -> None: ...
def overload2(func: Callable[..., Any]) -> None: ...

def cb(*, x: int) -> None: ...

overload1(cb)  # Error
overload2(cb)
```

**Comment by Kajiih** [comment]
Thank you.
Indeed, there is the same error when we remove the `*nested_dicts` parameters altogether. But another thing that's confusing here is that the behavior changes when the type of `*nested_dicts` changes in the example. Specifically, it is not considered fully overlapping in the second case.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.390.
